OsmAnd-java/src/main/java/net/osmand/router/GpxMultiSegmentsApproximation.java

```diff
@@ -119,15 +123,24 @@ private void addSegmentInternal(RouteSegmentAppr last, RouteSegment sg, List<Rou
 			System.out.println("** " + sg + " - not accepted");
 		}
 	}
-	
-	
+
 	private void addSegment(RouteSegmentAppr last, RouteSegment sg, List<RouteSegmentAppr> connected) {
 		if (sg == null) {
 			return;
 		}
-		if (sg.getRoad().getId() != last.segment.road.getId() || 
-				Math.min(sg.getSegmentStart(), sg.getSegmentEnd()) != 
-				Math.min(last.segment.getSegmentStart(), last.segment.getSegmentEnd())) {
+
+		int prev = last.gpxStart;
+		int next = Math.min(last.gpxNext(), gpxPoints.size() - 1);
```

Comment:
too complicated , definitely don't like this loops vs simple existing code doesn't make sense

Reply:
the `loops` was your code initially:

```
		if (sg.getRoad().getId() != last.segment.road.getId() ||
				sg.getSegmentStart() != last.segment.getSegmentStart() // 1. allow loops for rec tracks
//				Math.min(sg.getSegmentStart(), sg.getSegmentEnd()) != 
//				Math.min(last.segment.getSegmentStart(), last.segment.getSegmentEnd()) // 2. don't allow loops
				) {
```

Reply:
created method to keep addSegment simple